### PR TITLE
Add more optionxform tests

### PR DESCRIPTION
**Pull request by abravalheri**
@FlorianWilhelm is this enough for #55 ?

### CODE REVIEW

**Comment by FlorianWilhelm** [comment]
Thanks, @abravalheri for the unit tests. I think the critical code paths are those when `.detach()` is called before, right? Isn't it that in this case resolution works different?

**Comment by abravalheri** [comment]
Hi @FlorianWilhelm, if we compare the current state of the code with v2.0 and focus on the `optionxform` usage, I think there are 2 main differences:

**a)** We are using a intermediate `Parser` object that gets its `optionxform` by the `ConfigUpdater`. Since all `ConfigUpdater` objects use `Parser` objects now, the existing tests ensure backwards compatibility and that nothing was broken
**b)** When transferring/copying options from one object to another, we now use the *target*'s `optionxform` instead of the *origin*'s one - this is done to avoid the chicken/egg problem with the `NotAttachedError`.

(I might be forgetting something here ...)

In the original proposal for the `__deepclone__`, I did change a lot the critical code paths when `.detach()` was called... but after our discussions I reverted the code to be closer to the original, so now it does not mess up too much with `optionxform`...

The tests I am introducing here focus in ensuring compatibility with the usage listed in stdlib ConfigParser docs and **b)**

---
That said, since yesterday, I did realise we were missing to one code path in `Section.__setitem__`: we were just checking name consistency (the key being set and the actual key of the passed option object) for pre-existing keys, not the new ones.

I my latest commit, I reorganised `Section.__setitem__` to un-nest it a little bit (now the branching depth is 2, previously it was 3) and to share the key consistency check between pre-existing and new keys.

**Comment by FlorianWilhelm** [comment]
Thanks, @abravalheri, cool, got it, then we are good to go :-) Last step towards 3.0 it seems.
